src/ray/common/memory_monitor.cc

```diff
@@ -111,41 +111,107 @@ std::tuple<int64_t, int64_t> MemoryMonitor::GetMemoryBytes() {
   return std::tuple(system_used_bytes, system_total_bytes);
 }
 
-int64_t MemoryMonitor::GetCGroupV1MemoryUsedBytes(const char *path) {
+int64_t MemoryMonitor::GetCGroupV1MemoryUsedBytes(const char *stat_path,
+                                                  const char *usage_path) {
+  // How does this function calculate in-used memory from cgroup memory info file?
+  // It reads 2 cgroup files:
+  //  mem stat file: /sys/fs/cgroup/memory/memory.stat
+  //  mem usage file: /sys/fs/cgroup/memory/memory.usage_in_bytes
+  // Formula:
+  //  OS_managed_cache_and_buffer = `memory.stat.total_cache - memory.stat.total_shmem`
+  //  used_memory = `memory.usage_in_bytes` - OS_managed_cache_and_buffer
+  //
+  // This value is consistent with values `MemTotal` `MemAvailable` `MemFree` in `/proc/meminfo`
+  //  and they have relationship of:
+  //  - `memory.usage_in_bytes` == `MemTotal` - `MemFree`
+  //  - `memory.stat.total_cache` == `MemAvailable` - `MemFree`
+  //  - OS_managed_cache_and_buffer = `memory.stat.total_cache` - `shmem`
+  //
+  // Explanation: What's this part `OS_managed_cache_and_buffer` memory for and why
+  //   we should treat this part memory as "Not-in-used" ?
+  //   Linux OS tries to fully use the whole physical memory size, so that in many
+  //   cases we can observe that the system has very little `MemFree` size,
+  //   but at the same time system might have a large `MemAvailable` size,
+  //   then this part `MemAvailable - MemFree` size is borrowed by Linux OS
+  //   to cache pages / buffers , BUT, once user process requests to allocate memory,
+  //   and there is no sufficient free memory, OS will evict cache data out of this
+  //   part memory and allocate them to user proces.
```

Comment:
NOTE:

Current master Ray code calculation counts "OS_managed_cache_and_buffer" as "in-used" memory, which causes the issue that it heavily underestimate available memory,
and it leads to Ray memory monitor misleadingly triggered to kill Ray tasks while there are lots of available memory.

This is a very big issue.

Because for many IO-heavy tasks, it can quickly fill up physical memory with "OS_managed_cache_and_buffer", then Ray monitor will misleadingly think OS has very few available memory and then kill Ray tasks.

Reply:
Question, do we want to reserve some memory for cache+buffer, maybe have a flag to define certain threshold? It may impact some performance if there is no memory cache.

Reply:
@lanbochen-anyscale

> do we want to reserve some memory for cache+buffer

OS managed "cache+buffer" cache isn't used in this way. Linux OS fully decides what to cache to "cache+buffer", you can't control it.

But you can set "RAY_memory_usage_threshold" to be 0.8 so that 20% memory are reserved for "cache+buffer".

For the function "get_cgroupv1_used_memory" I modified, it should return the correct result of "in-used" memory, if you return some value like "in-used-memory + reserved_size", it makes code logic in a mess. You can consider to add logic for "reserved_size", but not in this function. For the "reserved_size" logic you mentioned, tuning "RAY_memory_usage_threshold" should be a better idea.

Reply:
More explanation:

> Where is "get_cgroupv1_used_memory" function used ?

1. It is used by "Ray memory monitor" to detect out-of-memory case in advance. i.e if `used_memory / physical_memory >  RAY_memory_usage_threshold` occurs, killing Ray tasks to prevent real OOM. Then if you change "used_memory" to be some value like "real_used_memory + reserved_size", it will heavily confuses user "what does the 'used_memory'" mean" ?  It is an an ugly patch containing disordered logic.

2. It is used in "Ray dashboard" to display available memory. This also requires it to return a "correct" value but not a tweaked value with some fuzzy "reserved_size" , otherwise user would ask why the dashboard shows wrong available memory size ?